Source/WebCore/rendering/RenderAttachment.cpp

```diff
@@ -79,6 +79,7 @@ void RenderAttachment::layout()
         RenderReplaced::layout();
         if (hasShadowContent())
             layoutShadowContent(intrinsicSize());
+        attachmentElement().requestWideLayoutIcon();
```

Comment:
`requestWideLayoutIcon()` is cheap to call multiple times, it only does a cross-process request when really necessary. Calling it during layout (as is done with legacy rendering) ensures that attributes have stabilized, otherwise there would be a risk of requesting an unnecessary icon.

But it may indeed be more appropriate to change this to an on-attribute-changed scheme, I'll think about it...

Reply:
Looking at the code again, I still think calling `requestWideLayoutIcon()` during painting may be the most efficient, it should only do one actual request in most cases, and none if the attachment is off-screen.
Otherwise we'd need to instrument at least `attributeChanged()`, but since we rely on two properties (`title` and `type`), that would mean sending two requests in almost all cases.

Reply:
Hmm yeah this is tricky. Generally we do try to avoid changing the DOM from rendering code (which is sort what this is doing by mutating the image src attribute), since it creates cycles:
1. DOM is initialized
2. rendering/layout
3. DOM changes due to rendering
4. rendering/layout
etc.

I know there's a check to prevent the infinite cycle, but it would be good to avoid this pattern if you can find a way, not super important though.

Reply:
In this case, it wouldn't be an infinite cycle:
1. DOM init, attributes are changed and settled.
2. Rendering/layout (with empty img), triggering the icon request.
3. "Later", the icon arrives, img data is changed once.
4. Rendering/layout, hopefully limited to that fixed-sized img area?
5. The end.

Note that the icon request is asynchronous, so even if we sent it before the first rendering, the icon could still arrive later and trigger a second rendering anyway.

Reply:
Also we need the `renderRect` of the img when requesting the icon.

Reply:
Yeah I know, `m_needsIconRequest` prevents this from being infinite, just saying the general "update DOM during rendering" pattern is something we try to avoid if it's low effort.

Reply:
One thing that may make me reconsider: Some ref tests seem a bit flaky, where I see either the actual or expected image is missing the icon! Moving the request a bit sooner _may_ help.

However, I'm not convinced this would be the right thing to do, since I think the real issue is that ref tests probably just take their snapshot at the end of the first full paint (right?), so as long as we have this asynchronous icon request, we'll always have the risk that the response could come too late and make the test fail; moving the request earlier would just make the failure less likely, not make it disappear completely.

So I would be more inclined in raising a separate radar about it, to solve this issue more broadly for all attachment tests (narrow/legacy and wide/modern). In case you know: Is there already a preferred way to delay reference tests taking their snapshot?

Reply:
> In case you know: Is there already a preferred way to delay reference tests taking their snapshot?

Yes, `class="reftest-wait"` on the `<html>` tag lets you do so, then removing that class whenever the page is ready to be screenshot.

Reply:
Thanks. "Whenever the page is ready" is not trivial to know in this case: The icon is hidden inside a custom renderer or a UA shadow tree. 😅
I'm looking at adding an early request...

Reply:
Added an early request when attributes are set; there's a bit of complexity to avoid doing multiple requests if multiple attributes are set in one operation. There's still a potential icon request during rendering, in the unexpected case where no attributes are set.

Reply:
FYI, I did a quick test (without this patch) : in "onload", looping `requestAnimationFrame` until a test attachment element received its icon, and it takes between 1 and 3 loops to get there! So I'm scared that all our attachment ref-tests are currently just comparing attachment bricks without icons 😱; intermittent failures probably happen when one side receives its icon unexpectedly quickly.
I'll prepare a proper bug separately... I don't think that should impact this patch here (as in, this icon-request race has been here for a long time, even before wide/modern attachments, so it shouldn't weigh in our decision to land this.)